plugin-server/src/main/ingestion-queues/batch-processing/each-batch.ts

```diff
@@ -31,42 +31,43 @@ export async function eachBatch(
         for (const messageBatch of messageBatches) {
             const batchSpan = transaction.startChild({ op: 'messageBatch', data: { batchLength: messageBatch.length } })
 
-            if (!isRunning() || isStale()) {
-                status.info('🚪', `Bailing out of a batch of ${batch.messages.length} events (${loggingKey})`, {
-                    isRunning: isRunning(),
-                    isStale: isStale(),
-                    msFromBatchStart: new Date().valueOf() - batchStartTimer.valueOf(),
-                })
-                await heartbeat()
-                return
-            }
+            const processingPromises = messageBatch.map((message: KafkaMessage) => eachMessage(message, queue))
+            processingPromises.push(heartbeat()) // Signal to the Kafka cluster that we are actively processing events
+            await Promise.all(processingPromises)
 
-            const lastBatchMessage = messageBatch[messageBatch.length - 1]
-            await Promise.all(messageBatch.map((message: KafkaMessage) => eachMessage(message, queue)))
+            batchSpan.finish()
+        }
 
-            // this if should never be false, but who can trust computers these days
-            if (lastBatchMessage) {
-                resolveOffset(lastBatchMessage.offset)
-            }
+        // Commit offsets once at the end of the batch. We run the risk of duplicates
+        // if the pod is prematurely killed in the middle of a batch, but this allows
+        // us to process events out of order within a batch, for higher throughput.
+        const commitSpan = transaction.startChild({ op: 'offsetCommit' })
+        const lastMessage = batch.messages.at(-1)
+        if (lastMessage) {
+            resolveOffset(lastMessage.offset)
```

Comment:
My understanding is resolveOffset is roughly equivalent of e.g. librdkafka store offset, is that right?

Reply:
Indeed, it passes the offset to commit for that partition, but cannot commit it, that's what the `commitOffsetsIfNecessary` below is trying to do.

This is [what the runner does](https://github.com/tulios/kafkajs/blob/ff3b1117f316d527ae170b550bc0f772614338e9/src/consumer/runner.js#L336-L338) if `eachBatchAutoResolve` is enabled. Because in this consumer we only handle one partition at a time, we don't need to pass the topic and partition params.